File: src/main/java/com/gitlab/daring/image/command/CommandRegistry.java
package com.gitlab.daring.image.command;

import com.gitlab.daring.image.command.combine.CombineCommands;
import com.gitlab.daring.image.command.drawing.DrawCommands;
import com.gitlab.daring.image.command.env.EnvCommands;
import com.gitlab.daring.image.command.feature.FeaturesCommands;
import com.gitlab.daring.image.command.structure.StructureCommands;
import com.gitlab.daring.image.command.transform.TransformCommands;
import com.gitlab.daring.image.common.BaseComponent;
import com.google.common.cache.Cache;
import com.typesafe.config.Config;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.gitlab.daring.image.MainContext.mainContext;
import static com.gitlab.daring.image.command.CommandUtils.parseArgs;
import static com.gitlab.daring.image.command.CommandUtils.splitScript;
import static com.gitlab.daring.image.util.CacheUtils.buildClosableCache;
import static com.gitlab.daring.image.util.ExceptionUtils.throwArgumentException;
import static com.gitlab.daring.image.util.ExtStringUtils.splitAndTrim;
import static java.util.Collections.emptyList;

public class CommandRegistry extends BaseComponent implements AutoCloseable {

    //TODO move to main context
    public static final CommandRegistry Instance = new CommandRegistry();

    final Config cmdConfig = getConfig("commands");
    final Map<String, Command.Factory> factories = new HashMap<>();
    final Cache<String, Command> cache = buildClosableCache(config.getString("cache"));

    public CommandRegistry() {
        super("isb.CommandRegistry");
        EnvCommands.register(this);
        TransformCommands.register(this);
        CombineCommands.register(this);
        DrawCommands.register(this);
        StructureCommands.register(this);
        FeaturesCommands.register(this);
        mainContext().closeEvent.onFire(this::close);
    }

    public void register(String name, Command.Factory f) {
        factories.put(name, f);
    }

    public ScriptCommand parseScript(String script) {
        List<Command> cmds = splitScript(script).map(this::getCommand).toList();
        return new ScriptCommand(script, cmds);
    }

    private Command getCommand(String cmdStr) {
        Command cmd = cache.getIfPresent(cmdStr);
        if (cmd == null) {
            cmd = parseCommand(cmdStr);
            if (cmd.isCacheable()) cache.put(cmdStr, cmd);
        }
        return cmd;
    }

    private Command parseCommand(String cmdStr) {
        List<String> ss = splitAndTrim(cmdStr, "()").toList();
        String name = ss.get(0);
        String[] args = parseArgs(ss.size() > 1 ? ss.get(1) : "", getDefArgs(name));
        Command.Factory cf = factories.get(name);
        if (cf == null) throwArgumentException("Invalid command " + cmdStr);
        return cf.create(args);
    }

    private List<String> getDefArgs(String name) {
        return cmdConfig.hasPath(name) ? cmdConfig.getStringList(name) : emptyList();
    }

    @Override
    public void close() {
        cache.invalidateAll();
        cache.cleanUp();
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/CommandRegistry.kt
package com.gitlab.daring.image.command

import com.gitlab.daring.image.MainContext
import com.gitlab.daring.image.command.CommandUtils.parseArgs
import com.gitlab.daring.image.command.CommandUtils.splitScript
import com.gitlab.daring.image.command.combine.CombineCommands
import com.gitlab.daring.image.command.drawing.DrawCommands
import com.gitlab.daring.image.command.env.EnvCommands
import com.gitlab.daring.image.command.feature.FeaturesCommands
import com.gitlab.daring.image.command.structure.StructureCommands
import com.gitlab.daring.image.command.template.TemplateCommands
import com.gitlab.daring.image.command.transform.TransformCommands
import com.gitlab.daring.image.config.ConfigUtils.defaultConfig
import com.gitlab.daring.image.util.CacheUtils.buildClosableCache
import com.gitlab.daring.image.util.ExceptionUtils.throwArgumentException
import com.gitlab.daring.image.util.ExtStringUtils.splitAndTrim
import java.util.*

typealias CommandFactory = (Array<String>) -> Command

class CommandRegistry : AutoCloseable {

    val config = defaultConfig.getConfig("isb.CommandRegistry")
    val factories = HashMap<String, CommandFactory>()
    val cache = buildClosableCache<String, Command>(config.getString("cache"))

    val commands get() = factories.keys

    init {
        EnvCommands.register(this)
        TransformCommands.register(this)
        CombineCommands.register(this)
        DrawCommands.register(this)
        StructureCommands.register(this)
        TemplateCommands.register(this)
        FeaturesCommands.register(this)
        MainContext.closeEvent.onFire { this.close() }
    }

    fun register(name: String, f: CommandFactory) {
        factories.put(name, f)
    }

    fun parseScript(script: String): ScriptCommand {
        val cmds = splitScript(script).map(this::getCommand)
        return ScriptCommand(script, cmds)
    }

    private fun getCommand(exp: String): Command {
        cache.getIfPresent(exp)?.let { return it }
        return parseCommand(exp).apply { if (cacheable) cache.put(exp, this) }
    }

    private fun parseCommand(cmdStr: String): Command {
        val ss = splitAndTrim(cmdStr, "()")
        val args = parseArgs(ss.getOrNull(1) ?: "")
        val cf = factories[ss[0]] ?: throwArgumentException("Invalid command " + cmdStr)
        return cf(args)
    }

    override fun close() {
        cache.invalidateAll()
        cache.cleanUp()
    }

}